### PR TITLE
[FLINK-26877] Introduce auto type inference for filter push down

### PR DESCRIPTION
**Pull request by LadyForest**
* Add PredicateConverterTest

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 1]
```diff
@@ -106,14 +108,20 @@ private Predicate visitBiFunction(
         Optional<Integer> field = extractFieldReference(children.get(0));
         Optional<Literal> literal;
         if (field.isPresent()) {
-            literal = extractLiteral(children.get(1));
+            literal =
+                    extractLiteral(
+                            ((FieldReferenceExpression) children.get(0)).getOutputDataType(),
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:113]: `extractFieldReference` returns `Optional< FieldReferenceExpression>`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 2]
```diff
@@ -131,22 +139,28 @@ private Predicate visitBiFunction(
...
+            LogicalType expectedLogicalType = expectedType.getLogicalType();
+            DataType actualType = valueExpression.getOutputDataType();
+            LogicalType actualLogicalType = actualType.getLogicalType();
+            DataStructureConverter<Object, Object> converter = getConverter(expectedType);
+            Object value = valueExpression.getValueAs(actualType.getConversionClass()).get();
+            if (!actualLogicalType.equals(expectedLogicalType)
+                    && supportsImplicitCast(actualLogicalType, expectedLogicalType)) {
+                try {
+                    value = TypeUtils.castFromString(value.toString(), expectedLogicalType);
+                } catch (UnsupportedOperationException e) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:150]: Let catch `Exception ignore`?
I don't think we want to throw an exception here no matter what the situation is.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 3]
```diff
@@ -131,22 +139,28 @@ private Predicate visitBiFunction(
...
-                                                    valueExpression
-                                                            .getValueAs(type.getConversionClass())
-                                                            .get())))
-                    : Optional.empty();
+            LogicalType expectedLogicalType = expectedType.getLogicalType();
+            DataType actualType = valueExpression.getOutputDataType();
+            LogicalType actualLogicalType = actualType.getLogicalType();
+            DataStructureConverter<Object, Object> converter = getConverter(expectedType);
+            Object value = valueExpression.getValueAs(actualType.getConversionClass()).get();
+            if (!actualLogicalType.equals(expectedLogicalType)
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:146]: Use `LogicalTypeRoot` equals.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 4]
```diff
@@ -103,50 +107,55 @@ private Predicate visitBiFunction(
...
+                    && supportsImplicitCast(actualLogicalType, expectedLogicalType)) {
+                try {
+                    value = TypeUtils.castFromString(value.toString(), expectedLogicalType);
+                } catch (UnsupportedOperationException e) {
+                    throw new UnsupportedExpression();
+                }
+            }
+            if (supportsPredicate(expectedLogicalType)) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:154]: we can return null first if `supportsPredicate` returns false

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 5]
```diff
@@ -103,50 +107,55 @@ private Predicate visitBiFunction(
...
-                    : Optional.empty();
+            LogicalType expectedLogicalType = expectedType.getLogicalType();
+            DataType actualType = valueExpression.getOutputDataType();
+            LogicalType actualLogicalType = actualType.getLogicalType();
+            DataStructureConverter<Object, Object> converter = getConverter(expectedType);
+            Object value = valueExpression.getValueAs(actualType.getConversionClass()).get();
+            if (!actualLogicalType.equals(expectedLogicalType)
+                    && supportsImplicitCast(actualLogicalType, expectedLogicalType)) {
+                try {
+                    value = TypeUtils.castFromString(value.toString(), expectedLogicalType);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:149]: `TypeUtils.castFromString` returns internal value. (Like `BinaryStringData`)

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java (hunk 2 above)
Comment (LadyForest) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:150]: > Let catch Exception ignore?
I don't think we want to throw an exception here no matter what the situation is.

the reason catching exception here is to let `PredicateTest#testUnsupportedExpression` pass. If not caught here, once the casting fails, it'll directly throw `UnsupportedOperationException`, other than `PredicateConverter.UnsupportedExpression`. I think it depends on whether we want an unified exception thrown by `PredicateConverter`

Comment (LadyForest) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java]: finally got ya

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 6]
```diff
@@ -103,50 +106,59 @@ private Predicate visitBiFunction(
...
-                                    type.getLogicalType(),
-                                    getConverter(type)
-                                            .toInternalOrNull(
-                                                    valueExpression
-                                                            .getValueAs(type.getConversionClass())
-                                                            .get())))
-                    : Optional.empty();
+            DataType actualType = valueExpression.getOutputDataType();
+            LogicalType actualLogicalType = actualType.getLogicalType();
+            Object value = valueExpression.getValueAs(actualType.getConversionClass()).get();
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:146]: get -> return Optional.empty.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**